providers/supermicro/x13.go

```diff
@@ -0,0 +1,271 @@
+		}, nil
+	default:
+		// ideally we never get in this position, since theres model number validation in parent callers.
+		return nil, errors.New("unsupported model for X13 BIOS fw install: " + c.model)
+	}
+}
+
+// redfish OEM fw install parameters
+func (c *x13) bmcFwInstallParams() map[string]bool {
+	return map[string]bool{
+		"PreserveCfg": true,
+		"PreserveSdr": true,
+		"PreserveSsl": true,
+		"BackupBMC":   false,
+	}
+}
+
+func (c *x13) redfishParameters(component, targetODataID string) (*rfw.RedfishUpdateServiceParameters, error) {
+	errUnsupported := errors.New("redfish parameters for x13 hardware component not supported: " + component)
+
+	oem := OEM{}
+
+	biosInstallParams, err := c.biosFwInstallParams()
+	if err != nil {
+		return nil, err
+	}
+
+	switch strings.ToUpper(component) {
+	case common.SlugBIOS:
+		oem.Supermicro.BIOS = biosInstallParams
+	case common.SlugBMC:
+		oem.Supermicro.BMC = c.bmcFwInstallParams()
+	default:
+		return nil, errUnsupported
+	}
+
+	b, err := json.Marshal(oem)
+	if err != nil {
+		return nil, errors.Wrap(err, "error preparing redfish parameters")
+	}
+
+	return &rfw.RedfishUpdateServiceParameters{
+		// NOTE:
+		// X13s support the OnReset Apply time for BIOS updates if we want to implement that in the future.
+		OperationApplyTime: constants.OnStartUpdateRequest,
+		Targets:            []string{targetODataID},
+		Oem:                b,
+	}, nil
+}
+
+func (c *x13) redfishOdataID(ctx context.Context, component string) (string, error) {
+	errUnsupported := errors.New("unable to return redfish OData ID for unsupported component: " + component)
+
+	switch strings.ToUpper(component) {
+	case common.SlugBMC:
+		return c.redfish.ManagerOdataID(ctx)
+	case common.SlugBIOS:
+		// hardcoded since SMCs without the DCMS license will throw license errors
+		return "/redfish/v1/Systems/1/Bios", nil
+		//return c.redfish.SystemsBIOSOdataID(ctx)
+	}
+
+	return "", errUnsupported
+}
+
+func (c *x13) firmwareInstallUploaded(ctx context.Context, component, uploadTaskID string) (installTaskID string, err error) {
+	if err = c.supportsInstall(component); err != nil {
+		return "", err
+	}
+
+	task, err := c.redfish.Task(ctx, uploadTaskID)
+	if err != nil {
+		e := fmt.Sprintf("error querying redfish tasks for firmware upload taskID: %s, err: %s", uploadTaskID, err.Error())
+		return "", errors.Wrap(brrs.ErrFirmwareVerifyTask, e)
+	}
+
+	taskInfo := fmt.Sprintf("id: %s, state: %s, status: %s", task.ID, task.TaskState, task.TaskStatus)
+
+	if task.TaskState != redfish.CompletedTaskState && task.TaskState != redfish.PendingTaskState {
+		return "", errors.Wrap(brrs.ErrFirmwareVerifyTask, taskInfo)
+	}
+
+	if task.TaskStatus != "OK" {
+		return "", errors.Wrap(brrs.ErrFirmwareVerifyTask, taskInfo)
+	}
+
+	err = c.redfish.StartUpdateForUploadedFirmwareNoTaskID(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	// X13s dont create a new task id when going from upload to install, so we pass through the same one
```

Comment:
Is this a redfish version thing? Or a model thing?

Reply:
Redfish.

X11 doesnt have a taskID at all, X12 creates a new taskID for each, and X13 keeps the same taskID